Support JITless Wasm import calls

danlliu opened the pull request:
#### 0b31e2848d7894a6a26fc22e04e24641ab426429
<pre>
Support JITless Wasm import calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=277830">https://bugs.webkit.org/show_bug.cgi?id=277830</a>
<a href="https://rdar.apple.com/133494023">rdar://133494023</a>

Reviewed by Yusuke Suzuki and Keith Miller.

This patch finishes JITless Wasm support by enabling JITless entry points for
imported functions, both from JS and Wasm. This allows us to run JetStream2.0
with JIT turned off.

* JSTests/wasm/stress/cc-infinite-int-glitch.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.result.i32.local.x.local.x.func.test2.export.string_appeared_here.param.x.i32.result.i32.result.i32.result.i32.local.x.local.x.local.x.async test):
* JSTests/wasm/stress/cc-int-to-int-to-js.js: Added.
(local.x.call.callee.callee):
(async test):
* JSTests/wasm/stress/wasm-js-call-many-return-types-on-stack-no-args.js:
(async test):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::generateWasmToJSStubs):
* Source/JavaScriptCore/wasm/WasmIndexOrName.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):

Canonical link: <a href="https://commits.webkit.org/282172@main">https://commits.webkit.org/282172@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7da5beb3878179f70b0d104f928a163a2db764f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14857 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12810 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13150 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50174 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8848 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38654 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53927 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30961 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35339 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11205 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11741 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55361 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11516 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67975 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61507 "Found 1 new JSC stress test failure: stress/scoped-arguments-table-should-be-tolerant-for-oom.js.bytecode-cache (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11259 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57551 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6235 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57766 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5134 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83271 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9382 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37419 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14607 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38503 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39599 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38248 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on JSTests/wasm/stress/cc-infinite-int-glitch.js:
```diff
@@ -21,6 +21,8 @@ async function test() {
     const { test, test2 } = instance.exports
 
+    print(test(5));
+    print(test2(5));
```
Let's remove them.

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
 
+# This is the interpreted analogue to WasmBinding.cpp:wasmToWasm
+op(wasm_to_wasm_wrapper_entry, macro()
+    loadq Callee-0x10[sp], ws1
```
Is `-0x10` the same in x86_64? Maybe it has difference since x86 already pushes return address to the stack?
Maybe it is prologueStackPointerDelta?

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -672,20 +672,21 @@ static inline UGPRPair doWasmCall(Register* partiallyConstructedCalleeFrame, JSW
+        calleeStackSlot = boxedCallee;
     }
 
-    Register& calleeStackSlot = partiallyConstructedCalleeFrame[static_cast<int>(CallFrameSlot::callee)];
-    ASSERT(calleeStackSlot.unboxedInt64() == 0xBEEF);
-    calleeStackSlot = boxedCallee;
```
Since boxed wasm callee information on the stack is necessary for stack scanning, maybe, we should keep it.
How about retrieving functionIndex from boxed callee?

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
+    jmp .done
+
+.funcindex:
+    noti ws1
```
I wonder if we can just get function index from boxed callee. You can unbox and load index from that.

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
+# This is the interpreted analogue to WasmBinding.cpp:wasmToJS
+op(wasm_to_js_wrapper_entry, macro()
+    # Load this before we create the stack frame, since we lose old cfr, which we wrote Callee to
+    loadq Callee-0x10[sp], ws1
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
+    const addressOfCalleeCalleeFromCallerPerspectiveBase = constexpr CallFrameSlot::callee * SlotSize - CallerFrameAndPCSize + PayloadOffset
+    storep ws0, addressOfCalleeCalleeFromCallerPerspectiveBase + CallerFrameAndPCSize - PrologueStackPointerDelta[sp]
+else
+    storei ws0, addressOfCalleeCalleeFromCallerPerspectiveBase + CallerFrameAndPCSize - PrologueStackPointerDelta + PayloadOffset[sp]
```
`addressOfCalleeCalleeFromCallerPerspectiveBase` is not defined?

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
+    # jit.storeWasmCalleeCallee(scratch, safeCast<int>(sizeof(CallerFrameAndPC)) - safeCast<int>(prologueStackPointerDelta()));
+if JSVALUE64
+    const addressOfCalleeCalleeFromCallerPerspectiveBase = constexpr CallFrameSlot::callee * SlotSize - CallerFrameAndPCSize + PayloadOffset
+    storep ws0, addressOfCalleeCalleeFromCallerPerspectiveBase + CallerFrameAndPCSize - PrologueStackPointerDelta[sp]
```
I think you can drop `CallerFrameAndPCSize` since you're just subtracting it then adding it again.

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -885,6 +884,262 @@ end
+    loadp JSWebAssemblyInstance::ImportFunctionInfo::boxedTargetCalleeLoadLocation[ws1], ws0
+    loadp [ws0], ws0
+
+    # jit.storeWasmCalleeCallee(scratch, safeCast<int>(sizeof(CallerFrameAndPC)) - safeCast<int>(prologueStackPointerDelta()));
```
Remove this comment I think it's clear from `addressOfCalleeCalleeFromCallerPerspectiveBase` we're storing the Wasm::Callee to the Callee slot.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -82,6 +82,7 @@ class Callee : public NativeCallee {
 
 private:
     const CompilationMode m_compilationMode;
+public:
```
Why is this public? There's already a getter.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIndexOrName.h:
```diff
@@ -102,18 +105,21 @@ struct IndexOrName {
     NameSection* nameSection() const { return m_nameSection.get(); }
 
-private:
+public:
```
ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -83,6 +85,7 @@ class Callee : public NativeCallee {
 private:
     const CompilationMode m_compilationMode;
     const IndexOrName m_indexOrName;
+    const size_t m_index;
```
Let's use `uint32_t`. Specifying the exact size is particularly good for fields used by JIT.
Also uint32_t should be sufficient given that wasm's function limit `WasmLimits.h`

Constellation replied:
Also use `load32` etc.

Constellation reviewed (approved):
r=me

danlliu committed:
Support JITless Wasm import calls
https://bugs.webkit.org/show_bug.cgi?id=277830
rdar://133494023

Reviewed by Yusuke Suzuki and Keith Miller.

This patch finishes JITless Wasm support by enabling JITless entry points for
imported functions, both from JS and Wasm. This allows us to run JetStream2.0
with JIT turned off.

* JSTests/wasm/stress/cc-infinite-int-glitch.js:
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.x.i32.result.i32.result.i32.local.x.local.x.func.test2.export.string_appeared_here.param.x.i32.result.i32.result.i32.result.i32.local.x.local.x.local.x.async test):
* JSTests/wasm/stress/cc-int-to-int-to-js.js: Added.
(local.x.call.callee.callee):
(async test):
* JSTests/wasm/stress/wasm-js-call-many-return-types-on-stack-no-args.js:
(async test):
* Source/JavaScriptCore/assembler/JITOperationList.cpp:
(JSC::llintOperations):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmEntryPlan.cpp:
(JSC::Wasm::EntryPlan::generateWasmToJSStubs):
* Source/JavaScriptCore/wasm/WasmIndexOrName.h:
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):

Canonical link: https://commits.webkit.org/282172@main
